### PR TITLE
Mempool and unconfirmed UTXO search optimization

### PR DESCRIPTION
**Pull request by ouziel-slama**
- Make mempool parsing optional
- New cache system for `getrawtransaction`
- Intialize cache at server start
- Raise error if cache not ready or mempool parsing disabled

### CODE REVIEW

File: counterpartylib/lib/blocks.py [hunk 1]
```diff
@@ -1024,6 +1028,11 @@ def follow(db):
             block_index += 1
 
         else:
+            if not config.PARSE_MEMPOOL:
+                db.wal_checkpoint(mode=apsw.SQLITE_CHECKPOINT_PASSIVE)
+                time.sleep(config.BACKEND_POLL_INTERVAL)
```
Comment (adamkrellenstein) [inline counterpartylib/lib/blocks.py]: These lines duplicate those at the end of this `else` block.

File: ChangeLog.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 * v9.51.1 (2015-04-??)
 	* Rename `server.api.log` to `server.access.log`
 	* Add `requests_timeout` parameter
+	* Add `rpc_batch_size` parameter and optimize RPC calls for block parsing and UTXO search
+	* Add `check_asset_conservation` parameter and optimize assets conservation checking
+	* Add `parse_mempool` parameter and disable mempool parsing by default
```
Comment (adamkrellenstein) [inline ChangeLog.md]: Can we just call this param `mempool`? `parse_mempool` sounds like it has something to do with *parsing*.

Comment (adamkrellenstein) [inline ChangeLog.md]: Or, perhaps `enable_mempool`?

File: counterpartylib/lib/backend/__init__.py [hunk 3]
```diff
@@ -111,10 +111,20 @@ def is_scriptpubkey_spendable(scriptpubkey_hex, source, multisig_inputs=False):
...
+
 def get_unspent_txouts(source, unconfirmed=False, multisig_inputs=False):
     """returns a list of unspent outputs for a specific address
     @return: A list of dicts, with each entry in the dict having the following keys:
     """
+    global MEMPOOL_CACHE_INITIALIZED
+
+    if unconfirmed and not config.PARSE_MEMPOOL:
+        raise MempoolError('Mempool parsing is not enabled; use `--parse-mempool` to enable it.')
+
```
Comment (adamkrellenstein) [inline counterpartylib/lib/backend/__init__.py]: Can we combine these two blocks under `if unconfirmed:`?

File: counterpartylib/lib/backend/__init__.py [hunk 4]
```diff
@@ -111,10 +111,20 @@ def is_scriptpubkey_spendable(scriptpubkey_hex, source, multisig_inputs=False):
...
+    pass
+
 def get_unspent_txouts(source, unconfirmed=False, multisig_inputs=False):
     """returns a list of unspent outputs for a specific address
     @return: A list of dicts, with each entry in the dict having the following keys:
     """
+    global MEMPOOL_CACHE_INITIALIZED
+
+    if unconfirmed and not config.PARSE_MEMPOOL:
+        raise MempoolError('Mempool parsing is not enabled; use `--parse-mempool` to enable it.')
```
Comment (adamkrellenstein) [inline counterpartylib/lib/backend/__init__.py]: `text = `Mempool is not enabled.'`

File: counterpartylib/lib/blocks.py (hunk 1 above)
Comment (ouziel-slama) [inline counterpartylib/lib/blocks.py]: I put all this block in a big `if`. I don't think it's worth here to make a function for this 2 lines.

File: ChangeLog.md (hunk 2 above)
Comment (ouziel-slama) [inline ChangeLog.md]: yep! done.

File: counterpartylib/lib/backend/__init__.py (hunk 3 above)
Comment (ouziel-slama) [inline counterpartylib/lib/backend/__init__.py]: yep! also done.

File: counterpartylib/lib/backend/__init__.py (hunk 4 above)
Comment (ouziel-slama) [inline counterpartylib/lib/backend/__init__.py]: done

**Pull request merged**